osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs

```diff
@@ -77,6 +78,14 @@ public void ChangeUserState(int userId, MultiplayerUserState newState)
             });
         }
 
+        public void ChangeUserBeatmapAvailability(int userId, BeatmapAvailability newBeatmapAvailability)
+        {
+            Debug.Assert(Room != null);
+
+            ((IMultiplayerClient)this).UserBeatmapAvailabilityChanged(userId, newBeatmapAvailability);
+            ChangeUserState(userId, MultiplayerUserState.Idle);
```

Comment:
this logic should probably be in `StatefulMultiplayerClient`, as a compound operation. the normal multi client should really be doing this the same way. there shouldn't be any half-and-half; either it's one server call (just `UserBeatmapAvailabilityChanged`, which automagically resets user state back to idle server-side) both in tests and in real usage, or two server calls (`UserBeatmapAvailabilityChanged` + `UserStateChanged`) in both.

Reply:
The way this is doing in the enclosing method is exactly like the server, propagate availability change and change user state afterward, I could make it such that the "change user state" to idle should be part of the client's job, but I'm not quite sure how sensible that would be (and also actually won't be possible because `IMultiplayerServer.ChangeState` is meant to guard against changing state while beatmap availability not set to `LocallyAvailable)

Reply:
if receiving an `UserBeatmapAvailabilityChanged` message from the server implicitly means that the user has reverted to idle, then `StatefulMultiplayerClient.UserBeatmapAvailabilityChanged` should be updating the user's state to idle, no? which is another shortcoming of merging the two messages.

Reply:
no `UserStateChanged` would also be invoked along the state set in `changeAndBroadcastUserState`, before `UserBeatmapAvailabilityChanged` is invoked.

Reply:
never mind, got confused by method naming in the test client